### ISSUE TITLE
Allow shareable configs to be globally installed

### ISSUE THREAD

**Issue by klebba** [stylelint/stylelint#1973]
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

This is a feature request for "Shareable Configs" which behave the same way that they do in ESLint:

http://eslint.org/docs/developer-guide/shareable-configs

In ESLint, Shareable Configs may may be globally installed (`npm install -g eslint-config-myrules`) and will be found automatically when running `eslint` if they are prefixed with `eslint-config-`

This allows a developer to configure their workstation with their organization's linter rules to ensure consistency across projects. Any project which extends `eslint-config-myrules` in its linter configuration will be opted into the organization's rules.

> Which rule, if any, is this issue related to?

None.

> What CSS is needed to reproduce this issue?

None.

> What stylelint configuration is needed to reproduce this issue?

None.

> Which version of stylelint are you using?

e.g. `7.3.1`

> How are you running stylelint: CLI, PostCSS plugin, Node API?

"CLI with `stylelint **/*.css`"

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?

No.

> What did you expect to happen?

My custom `styleling-config-myrules` would not require local installation in every project.

> What actually happened (e.g. what warnings or errors you are getting)?

``` shell
Could not find "stylelint-config-myrules". Do you need a `configBasedir`?
```

I'm reporting this as a feature request only. I will not be able to implement it or suggest how it should be implemented. My desire is feature parity with ESLint, which seems to have a reasonable solution to this problem.

Thanks in advance!

**Comment by davidtheclark** [comment]
Thanks @klebba for the suggestion.

stylelint uses [cosmiconfig](https://github.com/davidtheclark/cosmiconfig), which searches up the file tree, but does not search `node_modules` for global installs. So if the global npm install is not on that route, it will not be found.

This is a change that should be made in cosmiconfig, then carried over to stylelint.

**Comment by klebba** [comment]
@davidtheclark thanks for considering.

**Comment by davidtheclark** [comment]
Changed my mind about ^^ — not sure this is a cosmiconfig issue or not. Probably not, actually, since it's about the resolution of `extends`.

**Comment by davidtheclark** [comment]
I've thought about this more and I'm not sure ... here are some reasons:

1. I think I recommend *against* what @klebba wants to do. I believe a better practice is to install the dependencies of each project *within that project*, instead of having projects that rely on global modules.
2. You can accomplish what's requested by providing an absolute path to your globally installed config package.
3. If we modify lookups in configs (extends, plugins, processors) so they read from both local and global modules, we create an invitation to works-on-my-machine bugs. If A has a plugin installed globally and B does not, A and B cannot share configs.

Does anybody else have opinions about this?

**Comment by klebba** [comment]
Thanks for summarizing @davidtheclark. My rationale for requesting this feature is to have parity with ESLint. As a user of both tools, it's frustrating when there are variations in the way they can be used. I think it's worth adding to your consideration that similar/popular libraries do allow this method for distributing config

**Comment by davidtheclark** [comment]
> My rationale for requesting this feature is to have parity with ESLint. 

Parity with ESLint is not a goal of this project. ESLint is a source of good ideas, one among many. For the reasons mentioned above, I'm not sure this is a great feature. Stylelint already suffers some from imitating certain ESLint patterns that turn out to have been less-than-ideal. So I think we need to discuss whether this is a good feature or not — regardless of ESLint.

**Comment by jeddy3** [comment]
> Stylelint already suffers some from imitating certain ESLint patterns that turn out to have been less-than-ideal

@klebba Hello. FYI, this is mainly in reference to adopting eslint's _array-based_ configuration object pattern, which we did without first weighing up the merits of such a pattern. This pattern has presented us with numerous issues since stylelint's inception. We don't want to make that same mistake again.

> I believe a better practice is to install the dependencies of each project within that project

Agreed. Mainly for the following reason:

A code base is likely to be compatible with a _particular version_ of an organisation's configuration, which is turn is likely to be compatible with a _particular version_ of stylelint. It seems _likely_ that an organisation's config will evolve over time (e.g. as new rules are added to stylelint), and it seems _unlikely_ that you'd want to roll these out across all (legacy) projects.

Using local dependencies allows you specify *what particular version of stylelint and your organisation's config* the code within a project should be linted against:

Project A's `package.json`:

```json
{ 
  "devDependencies": {
    "stylelint": "~7.2.0",
    "stylelint-config-my-org": "^1.0.0"
  }
}
```

Project B's `package.json`:

```json
{ 
  "devDependencies": {
    "stylelint": "~7.4.0",
    "stylelint-config-my-org": "^2.0.0"
  }
}
```

Where project B is newer, and as such uses a later version of stylelint and a later version of `stylelint-config-my-org`, which makes use of a new rule introduced between `stylelint@7.2` and `stylelint@7.4`.

@klebba Have you not run into this issue with your projects when using a global organisational eslint config?

I think the recommendation would be to change how you use ESlint configs to use local dependencies for the reasons outlined above. If you do this both stylelint and eslint can be configured in the same way, and you'll get the benefits outlined above across both tools.

Having said that, I'm keen to learn more about how you distribute your eslint configs within your org, so that we can be sure we end-up with the most appropriate solution to this issue :)

**Comment by klebba** [comment]
Hey @jeddy3 -- thanks for following up. I do understand the rationale for locally installed rules, but there are situations where globally installed rules are desirable too. Ideally stylelint would support both use cases, and leave it up to users to choose (as ESLint has done).

> A code base is likely to be compatible with a particular version of an organisation's configuration, which is turn is likely to be compatible with a particular version of stylelint. It seems likely that an organisation's config will evolve over time (e.g. as new rules are added to stylelint), and it seems unlikely that you'd want to roll these out across all (legacy) projects.

Sure, this is likely for some situations, but not all. Your definition of a "project" may not be what our definition is. A single git repository is not necessarily a project with distinct linter rules, it may only consist of a shared module used across projects. In projects consisting of many modules, locally installed linter rules are a chore to maintain. This pattern is common in JavaScript modules (hence the ESLint support), but also emerging in WebComponents. Take the Polymer Elements catalog for example:

https://elements.polymer-project.org

Or Atom packages:

https://atom.io/packages/

In my case, our catalog of distributable modules must always pass the JS and CSS linter rules we have defined as a team. Managing/synchronizing locally installed versions of our linter rules across many repositories is overhead we can skip by treating the rules as environment configuration, and not a per-repo thing. That's not to say we don't also use locally installed rules sometimes -- we use both installation styles with ESLint.

> @klebba Have you not run into this issue with your projects when using a global organisational eslint config?

No. After a year of using ESLint this way we haven't had any headaches using this strategy. I don't think we are alone:

https://github.com/airbnb/javascript/tree/master/packages/eslint-config-airbnb

Everyone in our organization must keep their linter rules up to date, and our CI forces new linter violations to be resolved before merging. It's super trivial to fix up a few offending lines once you have stabilized a set of rules.

I should clarify that my feature request is not necessarily because I desire parity with ESLint, more that the feature implemented in ESLint covers/informs my use case and parity would be ideal (so we could copy an existing pattern that works for us).

Further my aim is to lint _all_ of the code I write on my workstation with a globally installed set of rules, but only when I opt-in using a local `.stylelintrc` files to extend our global set:

```
{
    "extends": "stylelint-config-org-rules",
}
```

With ESLint this is easily accomplished, local and global installation "just works," and IDE integration is trivial. This is not the case with stylelint.

**Comment by davidtheclark** [comment]
Note that this is still possible and pretty simple with stylelint, you just need to provide an absolute path in `extends` instead of a module name.

However, it's possible that we could build this in. We'd need to think through the details of global vs. local implementation? @klebba since you are a user who is used to this and wants this, can you help us out? 

Here's the specific question I'm thinking of — Would it acceptable to you to have this restriction?

If a config is loaded from a global npm install, any paths within it (plugins, processors, extends, etc.) will only work if they also point to global npm packages. Conversely, any config that is *not* loaded from a global npm install will not look global npm packages for any of its paths.

**Comment by klebba** [comment]
We can't commit an absolute path that will work across environments and workstations, so this isn't a solution for us.

To be honest I'm not exactly sure how the ESLint config behaves with regard local/global config resolution restrictions. I would have to dig into their implementation to form an opinion about this. In our particular case we either use global config/plugin/etc or local config/plugin/etc and never mix.

**Comment by davidtheclark** [comment]
> We can't commit an absolute path that will work across environments and workstations, so this isn't a solution for us.

@klebba: You can do this by whatever means stylelint *would* do it if the feature were built-in. Is there an npm module to help find the global installation directory? Or is there a particular way to do it reliably?

>  I would have to dig into their implementation to form an opinion about this

Yes, somebody will have to do this.

I'm open to a PR implementing this. @stylelint/core anybody object?

**Comment by evilebottnawi** [comment]
@davidtheclark sgtm

**Comment by jeddy3** [comment]
> I'm open to a PR implementing this

SGTM.

@klebba Thank you for [writing up](https://github.com/stylelint/stylelint/issues/1973#issuecomment-254288683) your use case. It really helped clarify your request :)

**Comment by klebba** [comment]
@davidtheclark @jeddy3 thanks for hearing me out. Regrettably I won't be able to develop this feature any time soon. I'd contribute to the bounty for anyone willing to take a stab

**Comment by jeddy3** [comment]
> Regrettably I won't be able to develop this feature any time soon. 

No worries. There's no rush. We'll leave this issue open with the "help wanted" label :)

**Comment by EvgenyOrekhov** [comment]
I use stylelint programmatically and I bumped into this issue.

I discovered that passing the name of the config that I extend to `require.resolve()` solves the issue:

```js
{"extends": require.resolve("stylelint-config-standard")}
```

In the case of using an array we would need to map it over `require.resolve()` like this:

```js
config.extends = config.extends.map(require.resolve);
stylelint.lint({
    code,
    codeFilename,
    config
});
```

I hope this info helps to fix the issue in the upstream.

**Comment by klebba** [comment]
@EvgenyOrekhov interesting workaround -- thanks for sharing!

**Comment by tdmalone** [comment]
Thanks @EvgenyOrekhov, that's working well for us to load from a global installation of Stylelint, but unfortunately only if the config is installed locally. Did you have luck using it to reference a global config?

> If a config is loaded from a global npm install, any paths within it (plugins, processors, extends, etc.) will only work if they also point to global npm packages. Conversely, any config that is not loaded from a global npm install will not look global npm packages for any of its paths.

This would be ideal. I think this is how ESLint does it. If you install and run globally, everything attempts to resolve to the global `node_modules`. If you install and run locally, everything attempts to resolve locally.

**Comment by tdmalone** [comment]
Actually I've worked out how to work around this - over the command line. If you `export NODE_PATH="$(npm root --global)"`, you can then do `stylelint --config "${NODE_PATH}"/stylelint-config-whatever`. If you don't depend on any other packages in your globally installed config, you can skip setting `NODE_PATH` and just use the correct path directly.

If you're using yarn, unfortunately it doesn't have an equivalent for the `npm root` command ([yet](https://github.com/yarnpkg/yarn/issues/2388)), so you might need to [detect a bit more](https://github.com/yarnpkg/yarn/issues/2049#issuecomment-263183768).

If you're not doing this over the command line, you _might_ be able to use a package such as [requireg](https://www.npmjs.com/package/requireg) but you might be out of luck if your global config has dependencies on other configs.

**Comment by EvgenyOrekhov** [comment]
@tdmalone I don't use stylelint-cli, I use [my own cli tool](https://github.com/EvgenyOrekhov/lints) that depends on stylelint and uses it programmatically.

The tool uses the `require.resolve()` trick, but it helps only when both stylelint and a shareable config are installed the same way:

| stylelint        | shareable config           | works?  |
| ------------- |-------------| -----|
| global      | global | true |
| local      | local      |   true |
| global | local      |    false |
| local | global      |    false |

**Comment by daKmoR** [comment]
hmmm any update on this? I just got been bitten by this...

I would like to have everything globally installed... stylelint and stylelint-config-my... but doesn't seem to work as it can't find it unless I give the absolute path.. but then again that is not possible as each team member may have a different global folder...

so does anyone has a solution for it?

**Comment by zanona** [comment]
Perhaps it would be interesting to look at `eslint` with `eslint-config-*` as a reference?

**Comment by ybiquitous** [comment]
I opened a PR! Please give me any feedbacks! 🙏
